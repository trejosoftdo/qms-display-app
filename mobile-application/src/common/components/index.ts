export { default as ActionButton } from './ActionButton';
export { default as ToConfigurationNavigationButton } from './ToConfigurationNavigationButton';
export { default as AppView } from './AppView';
export { default as Value } from './Value';
export { default as ConditionalContainer } from './ConditionalContainer';
export { default as Cards } from './Cards';
export { default as AppHeader } from './AppHeader';
