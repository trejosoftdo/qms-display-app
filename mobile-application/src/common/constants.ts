export const AUTHORIZATION_PENDING_CODE = 'authorization_pending';

export const VALIDATION_ERROR_STATUS_CODE = 400;

export const DEVICE_NOT_CONNECTED_ERROR_MESSAGE = 'Device is not connected';

export const CONNECTION_EXPIRED_ERROR_MESSAGE = 'Connection expired!';

export const APP_PROTO = 'app://';

export const DEVICE_ID_KEY = 'SECURE_DEVICE_ID';

export const CONNECTION_DETAILS_KEY = 'CONNECTION_DETAILS';

export const WEB_PLATFORM = 'web';

export const BEARER_PORTION = 'Bearer';

export const EMPTY_VALUE = '';

export const DEFAULT_REGULAR_FONT_NAME = 'Montserrat-Regular';

export const DEFAULT_LIGHT_FONT_NAME = 'Montserrat-Light';

export const DEFAULT_MEDIUM_FONT_NAME = 'Montserrat-Medium';

export const NORMAL_FONT_WEIGHT = 'normal';

export const ONE_SECOND_MILISECONDS = 1000;

export const INTERVAL_TIME = 10 * ONE_SECOND_MILISECONDS;

export const SERVICES_PATH = '/services';

export const TICKET_DETAILS_PATH = '/ticket-details';

export const HOME_PATH = '/';

export const CATEGORIES_PATH = '/categories';

export const CONFIGURATION_PATH = '/configuration';

export const NO_OP = () => {};


export const LANGUAGES = {
  SPANISH: 'es',
  ENGLISH: 'en',
};

export const CONFIG_OPTIONS = {
  INFORMATION: 'information',
  CONNECTION: 'connection',
  LANGUAGE: 'language',
};
