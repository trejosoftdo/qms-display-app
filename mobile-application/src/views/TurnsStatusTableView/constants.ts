export const BEING_ATTENDED_STATUS_CODE = "BEING_ATTENDED";

export const TO_BE_ATTENDED_STATUS_CODE = "TO_BE_ATTENDED";