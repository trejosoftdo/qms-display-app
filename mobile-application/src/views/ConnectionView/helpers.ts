import { ConnectionData, ConnectionDetails } from "../../common/models";
import { TRANSLATION_CONNECTED_MESSAGE_KEY, TRANSLATION_INITIAL_CONNECT_MESSAGE_KEY, TRANSLATION_SUCCESS_CONNECTION_MESSAGE_KEY, TRANSLATION_TO_CONNECT_MESSAGE_KEY } from "../../common/translations/translation-keys";

/**
 * Gets the correct message key from the data state
 * 
 * @param  {ConnectionData} data
 * @returns string
 */
export const getMessageKey = (data: ConnectionData, connectionDetails?: ConnectionDetails): string => {
  if (connectionDetails?.accessToken) {
    return TRANSLATION_CONNECTED_MESSAGE_KEY;
  }

  if (data?.tokens?.accessToken) {
    return TRANSLATION_SUCCESS_CONNECTION_MESSAGE_KEY;
  }

  return data?.userCode ? TRANSLATION_TO_CONNECT_MESSAGE_KEY : TRANSLATION_INITIAL_CONNECT_MESSAGE_KEY;
};
