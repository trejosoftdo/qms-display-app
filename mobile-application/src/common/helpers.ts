import { Platform } from 'react-native';
import { router } from 'expo-router';
import { APP_PROTO, EMPTY_VALUE, ONE_SECOND_MILISECONDS, WEB_PLATFORM } from './constants';

/** 
 * Checks if the platform is web
 * @returns boolean
 */
export const isWeb = (): boolean => Platform.OS === WEB_PLATFORM;

/** 
 * Gets the current time
 * @returns number
 */
export const getCurrentTime = (): number => new Date().getTime();

/**
 * Calculates the expire time
 * 
 * @param  {number} expiresIn
 * @returns number
 */
export const calculateExpireTime = (expiresIn: number): number => getCurrentTime() + expiresIn * ONE_SECOND_MILISECONDS;


/** 
 * Extracs icon name
 * @returns string
 */
export const extractIconName = (iconUrl: string): string => iconUrl.replace(APP_PROTO, EMPTY_VALUE)

/**
 * Navigates to a given path
 * 
 * @param  {string} pathname
 * @param  {any} params
 * @returns void
 */
export const goToPath = (pathname: string, params: Record<string, any> = undefined): void => {
  router.push({
    pathname,
    params,
  });
};

/**
 * Waits for certain time
 * @param  {number} timeout 
 * @returns Promise<any>
 */
export const delay = (timeout: number) => new Promise((resolve) => {
  setTimeout(resolve, timeout);
});

/**
 * Gets the response blob url
 * @param  {Response} response 
 * @returns Promise<string>
 */
export const getResponseBlobUrl = async (response: Response): Promise<string> => {
  const stream = await response.blob();
  return URL.createObjectURL(stream);
};
