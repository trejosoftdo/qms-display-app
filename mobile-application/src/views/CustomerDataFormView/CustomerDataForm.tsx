import * as React from 'react';
import { StyleSheet } from 'react-native';
import { Card, TextInput } from 'react-native-paper';
import { useTranslation } from 'react-i18next';
import { EMPTY_VALUE } from '../../common/constants';
import { TRANSLATION_CONTINUE_KEY, TRANSLATION_NAME_KEY } from '../../common/translations/translation-keys';
import ActionButton from '../../common/components/ActionButton';
import AppLightTheme from '../../common/theme';

/**
 * CustomerDataForm Props defines the props for the Customer Data Form Component.
 *
 * @property onSubmit - form on submit handler
 */
interface CustomerDataFormProps {
  onSubmit: (name: string) => void;
}

/**
 * A component a customer data form
 *
 * @param {CustomerDataFormProps} props - The props for the Customer Data Form component.
 */
const CustomerDataForm: React.FC<CustomerDataFormProps> = (props: CustomerDataFormProps) => {
  const { t } = useTranslation();
  const [name, setName] = React.useState(EMPTY_VALUE);
  return (
    <Card style={styles.container} mode="outlined">
      <Card.Content>
        <TextInput
          style={styles.input}
          label={t(TRANSLATION_NAME_KEY)}
          value={name}
          onChangeText={text => {
            setName(text);
          }}
          mode="flat"
          left={<TextInput.Icon icon="account" />}
        />
      </Card.Content>
      <Card.Actions style={styles.action}>
        <ActionButton
          icon="page-next"
          message={t(TRANSLATION_CONTINUE_KEY)}
          onPress={() => {
            props.onSubmit(name);
            setName(EMPTY_VALUE);
          }}
        />
      </Card.Actions>
    </Card>
  );
};

const styles = StyleSheet.create({
  container: {
    width: '100%',
    maxWidth: 500,
    backgroundColor: AppLightTheme.colors.background,
  },
  action: {
    marginBottom: 10,
    marginTop: 10,
    paddingRight: 16,
  },
  input: {
    backgroundColor: AppLightTheme.colors.background,
  }
});


export default CustomerDataForm;