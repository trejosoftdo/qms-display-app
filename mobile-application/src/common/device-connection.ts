import { v4 as uuidv4 } from 'uuid';
import { AuthHeaders, ConnectionDetails } from './models';
import { CONNECTION_DETAILS_KEY, DEVICE_ID_KEY } from './constants';
import { DEVICE_NOT_CONNECTED_ERROR, CONNECTION_EXPIRED_ERROR } from './errors';
import { getValue, setValue } from './store';
import { getNewAccessToken } from './services/auth';
import { BEARER_PORTION } from './constants';
import { getCurrentTime } from './helpers';


/**
 * Checks if a time has expired
 * 
 * @param  {{expiresAt:number}} {expiresAt}
 * @returns boolean
 */
const hasExpired = ({ expiresAt }: { expiresAt: number }): boolean => getCurrentTime() > expiresAt;

/**
 * Gets the device/installation id
 * @returns Promise<string>
 */
export const getDeviceId = async (): Promise<string> => {
  const fetchUUID = await getValue(DEVICE_ID_KEY);
  const deviceId = fetchUUID || uuidv4();
  await setValue(DEVICE_ID_KEY, deviceId.toString());
  return deviceId;
};

/**
 * Saves the connection details
 * 
 * @param  {ConnectionDetails} connectionDetails
 * @returns Promise<void>
 */
export const saveConnectionDetails = async (connectionDetails: ConnectionDetails): Promise<void> => {
  await setValue(CONNECTION_DETAILS_KEY, JSON.stringify(connectionDetails));
};

/**
 * Gets the stored connection details
 * 
 * @returns Promise<ConnectionDetails | undefined>
 */
export const getConnectionDetails = async (): Promise<ConnectionDetails | undefined> => {
  try {
    const data = await getValue(CONNECTION_DETAILS_KEY);
    const details = JSON.parse(data);
    return details as ConnectionDetails;
  } catch (error) {
    return undefined;
  }
};

/**
 * Gets the device auth headers to consume the APIs
 * 
 * @returns AuthHeaders
 */
export const getDeviceAuthHeaders = async (): Promise<AuthHeaders> => {
  const connectionDetails = await getConnectionDetails();

  if (!(connectionDetails?.applicationId && connectionDetails?.accessToken)) {
    throw DEVICE_NOT_CONNECTED_ERROR;
  }

  if (hasExpired(connectionDetails.accessToken)) {
    if (hasExpired(connectionDetails.refreshToken)) {
      throw CONNECTION_EXPIRED_ERROR;
    }

    connectionDetails.accessToken = await getNewAccessToken(
      connectionDetails.applicationId,
      connectionDetails.refreshToken.value,
    );
    await saveConnectionDetails(connectionDetails);
  }

  return {
    applicationId: connectionDetails.applicationId,
    authorization: `${BEARER_PORTION} ${connectionDetails.accessToken.value}`,
  };
};
