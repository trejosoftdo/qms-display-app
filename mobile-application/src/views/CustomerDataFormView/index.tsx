
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { TRANSLATION_CUSTOMER_DATA_FORM_MESSAGE_KEY } from '../../common/translations/translation-keys';
import { goToPath } from '../../common/helpers';
import { AppView } from '../../common/components';
import { CATEGORIES_PATH } from '../../common/constants';
import CustomerDataForm from './CustomerDataForm';

/**
 * CustomerDataFormViewProps defines the props for the Customer Data Form View Component.
 */
interface CustomerDataFormViewProps {}

/**
 * A component for the connect view
 *
 * @param {CustomerDataFormViewProps} props - The props for the Customer Data Form View component.
 */
const CustomerDataFormView: React.FC<CustomerDataFormViewProps> = (props: CustomerDataFormViewProps) => {
  const { t } = useTranslation();
  return (
    <AppView headerMessage={t(TRANSLATION_CUSTOMER_DATA_FORM_MESSAGE_KEY)}>
      <CustomerDataForm
        onSubmit={(customerName) => {
          goToPath(CATEGORIES_PATH, { customerName });
        }}
      />
    </AppView>
  );
};

export default CustomerDataFormView;