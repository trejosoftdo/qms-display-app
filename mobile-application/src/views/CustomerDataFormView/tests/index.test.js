import React from 'react';
import { act } from '@testing-library/react-hooks';
import { render, fireEvent, waitFor } from '@testing-library/react-native';
import { CATEGORIES_PATH } from '../../../common/constants';
import { goToPath } from '../../../common/helpers';
import CustomerDataFormView from '../';


jest.mock('../../../common/helpers');

describe('CustomerDataFormView component', () => {
  const mockCustomerName = 'mock-customer-name';

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders the component as expected', () => {
    expect(renderShallow(<CustomerDataFormView />)).toMatchSnapshot();
    expect(goToPath).toHaveBeenCalledTimes(0);
  });

  it('changes the language to the selected option', async () => {
    const { getByTestId } = render(<CustomerDataFormView />);
    
    const input = getByTestId('text-input-flat');
    const submitButton = getByTestId('button');
    
    await act(async () => {
      fireEvent.changeText(input, mockCustomerName);

      await waitFor(() => {
        expect(input.props.value).toEqual(mockCustomerName);
      });

      fireEvent.press(submitButton);
    });

    expect(goToPath).toHaveBeenCalledTimes(1);
    expect(goToPath).toHaveBeenCalledWith(CATEGORIES_PATH, { customerName: mockCustomerName });
  });
});
