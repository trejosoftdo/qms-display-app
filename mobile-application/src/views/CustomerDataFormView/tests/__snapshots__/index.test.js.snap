// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerDataFormView component renders the component as expected 1`] = `
<AppView
  headerMessage="Translated[translation:customerDataFormMessage]({})"
>
  <CustomerDataForm
    onSubmit={[Function]}
  />
</AppView>
`;
